File: DigitalImageProcessing/cpp/compute_affine_transform_matrix.cpp
/**
 * @File    : compute_affine_transform_matrix.cpp
 * @Brief   : 已知原始坐标和变换后的坐标，计算仿射变换矩阵
 * @Author  : Wei Li
 * @Date    : 2021-09-12
*/

#include <iostream>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>

// C++ API of OpenCV 输出参数有两种方式
// 1. 原始坐标和对应的变换后的坐标以 Point2f 进行存储
// 2. 原始坐标和对应的变换后的坐标以 Mat 进行存储
// -------------------------------------------
// 3. 利用 Linux OS 对 OpenCV 源代码进行编译和链接，同时利用 CMake 进行构建
// 编译和链接参考：https://gitee.com/weili_yzzcq/C-and-C-plus-plus/tree/master/OpenCV_Linux_Ubuntu
// 编译和链接参考：https://github.com/2694048168/C-and-C-plus-plus/tree/master/OpenCV_Linux_Ubuntu
// -------------------------------------------
int main(int argc, char** argv)
{
    // 1. 第一种方式
    cv::Point2f src[] = {cv::Point2f(0, 0), cv::Point2f(200, 0), cv::Point2f(0, 200)};
    cv::Point2f dst[] = {cv::Point2f(0, 0), cv::Point2f(100, 0), cv::Point2f(0, 100)};
    cv::Mat affine_transform_matrix = cv::getAffineTransform(src, dst);
    std::cout << affine_transform_matrix << std::endl;

    // 2. 第二种方式
    cv::Mat src_mat = (cv::Mat_<float>(3, 2) << 0, 0, 200, 0, 0, 200);
    cv::Mat dst_mat = (cv::Mat_<float>(3, 2) << 0, 0, 100, 0, 0, 100);
    cv::Mat affine_transform_mat = cv::getAffineTransform(src_mat, dst_mat);
    std::cout << affine_transform_mat << std::endl;

    // 通过基本的仿射变换进行矩阵乘法求解仿射变换矩阵
    // 缩放矩阵
    cv::Mat scale = (cv::Mat_<float>(3, 3) << 0.5, 0, 0, 0, 0.5, 0, 0, 0, 1);
    // 平移矩阵
    cv::Mat translation = (cv::Mat_<float>(3, 3) << 1, 0, 100, 0, 1, 200, 0, 0, 1);
    cv::Mat affine_transform;
    // 注意计算方向(先缩放后平移), 仿射矩阵计算和需要操作的过程相反！！！
    // cv::gemm(scale, translation, 1.0, cv::Mat(), 0, affine_transform, 0);
    cv::gemm(translation, scale, 1.0, cv::Mat(), 0, affine_transform, 0);
    std::cout << affine_transform << std::endl;

    // 等比例进行缩放 平移 旋转操作
    // center 变换中心的坐标；angle : 等比例缩放的系数  scale : 逆时针旋转的角度，单位为角度，而不是弧度；该值为负数，即为顺时针旋转。
    cv::Mat rotataion_matrix = cv::getRotationMatrix2D(cv::Point2f(40, 50), 30, 0.5);
    std::cout << rotataion_matrix << std::endl;
    
    return 0;
}

File: DigitalImageProcessing/python/compute_affine_transform_matrix.py
#!/usr/bin/env python3
# encoding: utf-8


"""
@Funciton: 已知原始坐标和变换后的坐标，计算仿射变换矩阵
@Python Version: 3.8
@Author: Wei Li
@Date: 2021-09-12
"""


import numpy as np
import cv2 as cv


def compute_affine_transform_matrix(src, dst):
    """利用 OpenCV 的 Python API 计算仿射变换矩阵。

    Args:
        src (float): 3 行 2 列的二维 ndarray，每一行代表一个原始坐标
        dst (float): 3 行 2 列的二维 ndarray，每一行代表一个变换后的坐标

    Returns:
        float: 返回仿射变换矩阵(3行3列)的前两行(2行3列)，
            最后一行为[0, 0, 1](齐次坐标表示法)
    """
    return cv.getAffineTransform(src=src, dst=dst)


def compute_affine_transform(scale, translation):
    """数字图像首先经过缩放变换，在进行平移变换。
    利用基本的仿射变换进行矩阵乘法求解出仿射变换矩阵。

    Args:
        scale (np.array): 基本的仿射变换——缩放
        translation (ndarray): 基本的仿射变换——平移
    
    Returns:
        float: 返回仿射变换矩阵(3行3列)
    """
    # 注意计算方向
    # return np.dot(scale, translation)
    return np.dot(translation, scale)


def compute_affine_transform_ratio(center, angle, scale):
    """等比例进行缩放并进行旋转。
    计算仿射变换矩阵，其本质还是计算各个基本的仿射矩阵进行相乘计算。

    Args:
        center ([type]): 变换中心的坐标
        angle ([type]): 等比例缩放的系数
        scale ([type]): 逆时针旋转的角度，单位为角度，而不是弧度；该值为负数，即为顺时针旋转。

    Returns:
        [type]: 通过旋转，平移，缩放等基本操作后的仿射变换矩阵(齐次坐标表示的前两行数值)
    """

    return cv.getRotationMatrix2D(center=center, angle=angle, scale=scale)


# --------------------------
if __name__ == "__main__":
    # 1. 方程方法
    src = np.array([[0, 0], [200, 0], [0, 200]], np.float32)
    dst = np.array([[0, 0], [100, 0], [0, 100]], np.float32)
    affine_transform_matrix = compute_affine_transform_matrix(src=src, dst=dst)
    print(f"The affine transform maxtrix type is\n {type(affine_transform_matrix)}")
    print(f"The affine transform maxtrix data type is\n {affine_transform_matrix.dtype}")
    print(f"The affine transform maxtrix is\n {affine_transform_matrix}")

    # 2. 矩阵方法
    scale = np.array([[0.5, 0, 0], [0, 0.5, 0], [0, 0, 1]])
    translation = np.array([[1, 0, 100], [0, 1, 200], [0, 0, 1]])
    affine_transform = compute_affine_transform(scale=scale, translation=translation)
    print(f"The affine transform maxtrix type is\n {type(affine_transform)}")
    print(f"The affine transform maxtrix data type is\n {affine_transform.dtype}")
    print(f"The affine transform maxtrix is\n {affine_transform}")

    # 3. 计算 (40, 50)为中心点进行逆时针旋转30度的仿射变换矩阵
    rotataion_matrix = compute_affine_transform_ratio(center=(40, 50), angle=30, scale=0.5)
    print(f"The affine transform maxtrix type is\n {type(rotataion_matrix)}")
    print(f"The affine transform maxtrix data type is\n {rotataion_matrix.dtype}")
    print(f"The affine transform maxtrix is\n {rotataion_matrix}")
